src/Google.Api.Gax.Grpc/BufferedClientWriter.cs

```diff
@@ -0,0 +1,285 @@
+                }
+            }
+        }
+
+        /// <summary>
+        /// Validates that we can write to the stream, optionally throwing if there's an error.
+        /// This is basically to avoid a big chunk of code appearing in WriteAsyncImpl.
+        /// </summary>
+        private bool ValidateStateForWrite(bool throwOnError)
+        {
+            if (_completed)
+            {
+                if (throwOnError)
+                {
+                    throw new InvalidOperationException($"Cannot call {nameof(WriteAsync)} after {nameof(CompleteAsync)}");
+                }
+                else
+                {
+                    return false;
+                }
+            }
+            if (_queue.Count == Capacity)
+            {
+                if (throwOnError)
+                {
+                    throw new InvalidOperationException($"Queue is full");
+                }
+                else
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Completes the stream when all buffered messages have been sent. This method can only be called
+        /// once, and further messages cannot be written after it has been called.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">This method has already been called.</exception>
+        /// <returns>A <see cref="Task"/> which will complete when the stream has finished being completed.</returns>
+        public Task CompleteAsync()
+        {
+            lock (_lock)
+            {
+                if (_completed)
+                {
+                    throw new InvalidOperationException($"Cannot call {nameof(CompleteAsync)} twice");
+                }
+                _completed = true;
+                if (_failedTask != null)
+                {
+                    return _failedTask;
+                }
+
+                var lastRequest = _queue.Last?.Value;
+                var request = new Request(null, null);
+                if (lastRequest == null)
+                {
+                    // If there's nothing in flight, send immediately. Otherwise,
+                    // we'll let HandleWriteComplete trigger sending us at the right time.
+                    request.WriteTo(_writer);
+                }
+                _queue.AddLast(request);
+                // No need to hook to HandleWriteComplete, as there are no more tasks to start.
+                return request.CompletionSource.Task;
+            }
+        }
+
+        private static void PropagateTaskResult(Task task, TaskCompletionSource<int> completionSource)
+        {
+            switch (task.Status)
+            {
+                case TaskStatus.Canceled:
+                    completionSource.TrySetCanceled();
+                    break;
+                case TaskStatus.Faulted:
+                    completionSource.SetException(task.Exception.InnerExceptions);
```

Comment:
Yes, it can be SetCanceled, which is better in terms of throwing if we've got a bug. Done.